retrieval/discovery/marathon.go

```diff
@@ -0,0 +1,177 @@
+)
+
+const (
+	// ImageLabel is the label that is used for the docker image running the service
+	ImageLabel clientmodel.LabelName = "image"
+)
+
+// MarathonDiscovery provides service discovery based on a Marathon instance
+type MarathonDiscovery struct {
+	servers        []string
+	name           string
+	useJobPrefix   bool
+	updateInterval time.Duration
+	done           chan struct{}
+	lastRefresh    map[string]*config.TargetGroup
+	client         marathon.AppListClient
+}
+
+// NewMarathonDiscovery creates a new Marathon based discovery
+func NewMarathonDiscovery(conf *config.MarathonSDConfig) *MarathonDiscovery {
+	return &MarathonDiscovery{
+		servers:        conf.Servers,
+		name:           conf.Name,
+		useJobPrefix:   conf.UseJobPrefix,
+		updateInterval: time.Duration(conf.UpdateInterval),
+		done:           make(chan struct{}),
+		client:         marathon.FetchMarathonApps,
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (md *MarathonDiscovery) Sources() []string {
+	var sources []string
+	tgroups, err := md.fetchTargetGroups()
+	if err == nil {
+		for source := range tgroups {
+			sources = append(sources, source)
+		}
+	}
+	return sources
+}
+
+// Run implements the TargetProvider interface.
+func (md *MarathonDiscovery) Run(ch chan<- *config.TargetGroup) {
+	defer func() {
+		log.Debug("Shutting down marathon discovery.")
+		close(ch)
+	}()
+
+	for {
+		select {
+		case <-md.done:
+			return
+		case <-time.After(md.updateInterval):
+			err := md.updateServices(ch)
+			if err != nil {
+				log.Errorf("Error while updating services: %s", err)
+			}
+		}
+	}
+}
+
+// Stop implements the TargetProvider interface.
+func (md *MarathonDiscovery) Stop() {
+	md.done <- struct{}{}
+}
+
+func (md *MarathonDiscovery) updateServices(ch chan<- *config.TargetGroup) error {
+	targetMap, err := md.fetchTargetGroups()
+	if err == nil {
+		// Update services which are still present
+		for _, tg := range targetMap {
+			ch <- tg
+		}
+
+		// Remove services which did disappear
+		for source := range md.lastRefresh {
+			_, ok := targetMap[source]
+			if !ok {
+				log.Debugf("Removing group for %s", source)
+				ch <- &config.TargetGroup{Source: source}
+			}
+		}
+
+		md.lastRefresh = targetMap
+	}
+	return err
+}
+
+func (md *MarathonDiscovery) fetchTargetGroups() (map[string]*config.TargetGroup, error) {
+	url := marathon.RandomAppsURL(md.servers)
+	apps, err := md.client(url)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := md.targetGroupsFromApps(apps)
+	return groups, nil
+}
+
+func (md *MarathonDiscovery) targetGroupsFromApps(apps *marathon.AppList) map[string]*config.TargetGroup {
+	tgroups := map[string]*config.TargetGroup{}
+	for _, a := range apps.Apps {
+		if md.isValidApp(&a) {
+			group := md.targetGroupFor(&a)
+			tgroups[group.Source] = group
+		}
+	}
+	return tgroups
+}
+
+func (md *MarathonDiscovery) targetGroupFor(app *marathon.App) *config.TargetGroup {
```

Comment:
`targetGroupForApp` is more consistent and self-descriptive without the method signature.